# Lecture du fichier de donnes

In [2]:
# importation des librairies
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os
# On a remplacer from pandas_profiling par from ydata_profiling
from ydata_profiling import ProfileReport
from pandas._libs import OutOfBoundsDatetime

In [56]:
# Lecture du fichier de données

data = pd.read_csv('Files/mouvements_comptables.txt', sep='|')

In [57]:
# Affichage des 5 premières lignes

data.head()

numero_compte                        lib_etendu  bilan   
0  ebd12c14b29d5e3c9c0785ea41787292  41ce64b36b9cfe92670dcdfa7ad4ca9b  21110  \
1  ebd12c14b29d5e3c9c0785ea41787292  fdbb71769ecf00d5778f657271bc9d9b  21110   
2  ebd12c14b29d5e3c9c0785ea41787292  ce2bf864bfcd82b4bf01120c79d7ba46  21110   
3  b9a371e2aa5b4a9a9801b639896db549  1fd98623f4623649bac0dd8c65ec49ce  21110   
4  a45c978dca989f17c7c8bb2cada9514f  234f3006bad9b99df4e4d08556484dc9  21110   

       dt_val      dt_ope  montant_signe signe  extourne  cpt_anal   
0  2022-01-28  2022-02-01          -99.0     -       1.0       NaN  \
1  2022-01-31  2022-02-01         -200.0     -       NaN       NaN   
2  2022-02-02  2022-02-01         1900.0     +       1.0       NaN   
3  2022-01-28  2022-02-01        -1000.0     -       1.0       NaN   
4  2022-01-28  2022-02-01        -2700.0     -       1.0       NaN   

    dt_tresor   dt_compta type_compta                          num_oper   
0  2022-02-01  2022-02-01           C  405834e2defca25e09dc43bb35baa8a9  \
1         NaN  2022-02-01         NaN  6c9af0676f860469c7c1789f13c33bbf   
2  2022-02-01  2022-02-01           C                               NaN   
3         NaN  2022-02-01           C  514a4b4d706abd42b192b9fa761894b6   
4         NaN  2022-02-01           C  ee3965f79a51f6aa2c2e6821fde47b99   

  cd_user cd_visa                   time  
0     NaN     NaN  2022-02-02 00:00:00.0  
1     NaN     NaN  2022-02-02 00:00:00.0  
2     NaN     NaN  2022-02-02 00:00:00.0  
3     NaN     NaN  2022-02-02 00:00:00.0  
4     NaN     NaN  2022-02-02 00:00:00.0

# On va commencer l'exploration des donnees

In [6]:
# On va utiliser pandas_profiling pour avoir un aperçu des données

profile = ProfileReport(data, title='Pandas Profiling Report', html={'style':{'full_width':True}})
# Enregistrer le rapport au format HTML
profile.to_file("Outputs/data_report.html")

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

# Conclusion :
A l'issue de la premiere analyse du rapport, on a 10000 observations dans notre dataset et 16 colonnes.
Nous avons 13 colonnes de type ctegoriciel, 2 de type numerique et 1 de type unsupported (il s'agit de la colone cpt_anal qui ne contient que des valeurs nulles).
Nous avons egalement 35470 cellules vides et 50 lignes dupliquees dans notre dataset.
Nous avons egalement pu obtenir un apercu rapide de la disribution des colonnes de notre dataset.
On a egalement des variables qui sont tres corelles, entre autres dt_tresor, cd_user et cd_visa qui sont tres corelles, mais cela est seulement du au fait que ces variables ont enormement de valeurs nulles

In [4]:
data.describe()

bilan  montant_signe  extourne  cpt_anal
count  10000.000000   9.999000e+03    8526.0       0.0
mean   25683.799900  -4.930480e+02       1.0       NaN
std    12702.380978   1.978045e+05       0.0       NaN
min    11110.000000  -4.689931e+06       1.0       NaN
25%    21110.000000  -1.505500e+03       1.0       NaN
50%    21110.000000  -9.000000e+01       1.0       NaN
75%    21210.000000   1.000000e+03       1.0       NaN
max    89490.000000   8.179528e+06       1.0       NaN

In [12]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 16 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   numero_compte  10000 non-null  object 
 1   lib_etendu     9999 non-null   object 
 2   bilan          10000 non-null  int64  
 3   dt_val         9986 non-null   object 
 4   dt_ope         9999 non-null   object 
 5   montant_signe  9999 non-null   float64
 6   signe          9999 non-null   object 
 7   extourne       8526 non-null   float64
 8   cpt_anal       0 non-null      float64
 9   dt_tresor      5521 non-null   object 
 10  dt_compta      9999 non-null   object 
 11  type_compta    8495 non-null   object 
 12  num_oper       8599 non-null   object 
 13  cd_user        2674 non-null   object 
 14  cd_visa        734 non-null    object 
 15  time           10000 non-null  object 
dtypes: float64(3), int64(1), object(12)
memory usage: 1.2+ MB


In [6]:
# On va etudier la colonne numero de compte

data['numero_compte'].value_counts()

numero_compte
c507b3ee99facefc0aaa42e5b40fe9ab    129
77d1c7cdb8262110ad7ec1b24f56195f     77
021a9a2d47e6e317191634fb783ec9a3     63
bc05265db32bc80f81520b91c1f6545f     58
6f93a70908558a8abbec920dabc3e29b     57
                                   ... 
66bb2bc5a622f158be5f89291552117c      1
fb59b527f94cecbffd7366d3b3076fe3      1
70a9f44b18d59250385f44222b868c7e      1
36d78f9bdc3d27c7b97281fb60291ed7      1
35c169d786f4ecbe7e4d72a09b64bb51      1
Name: count, Length: 5125, dtype: int64

On a 5125 numero de compte different dans notre dataset dont certains se repetent plusieurs fois.

In [20]:
# On va etudier la colonne libelle

data['lib_etendu'].value_counts()

lib_etendu
f63c78b6318dce39f42161d6c56df3e5    551
fdbb71769ecf00d5778f657271bc9d9b    260
5f9d3ba91a169b3787845d22e2600c4a    196
a5bff291014832114aa89d40ee8be7e2    169
dc264c7cb3824c5a0ae86d0caa1042e1    139
                                   ... 
6878ea83474a018b7258489918021ceb      1
5e3e78d9a6263b25ff9eec3625b39eba      1
a0458a07b5bac49a85c65251adb9e5e8      1
17933eaa7a81a0a750223b41a6e46e5d      1
feb34656f4f2399b5eb36094f71ed0dd      1
Name: count, Length: 5271, dtype: int64

In [21]:
data.lib_etendu.isnull().sum()

1

In [27]:
# On va voir la ligne qui contient la valeur nulle
data[data.lib_etendu.isnull()]

numero_compte lib_etendu  bilan dt_val dt_ope   
9199  5c01e4e913a1ba29ed1c2569e4373553        NaN  21110    NaN    NaN  \

      montant_signe signe  extourne  cpt_anal dt_tresor dt_compta type_compta   
9199            NaN   NaN       NaN       NaN       NaN       NaN         NaN  \

     num_oper cd_user cd_visa                   time  
9199      NaN     NaN     NaN  2022-02-02 00:00:00.0

In [7]:
# Voyons le nombre de transactions effectuees par la personne qui a le numero de compte

data[data.numero_compte == '5c01e4e913a1ba29ed1c2569e4373553']

numero_compte                        lib_etendu   
9199  5c01e4e913a1ba29ed1c2569e4373553                               NaN  \
9200  5c01e4e913a1ba29ed1c2569e4373553  9b168d19d3633aeff57ad22a0b79cbbf   
9201  5c01e4e913a1ba29ed1c2569e4373553  f63c78b6318dce39f42161d6c56df3e5   

      bilan      dt_val      dt_ope  montant_signe signe  extourne   
9199  21110         NaN         NaN            NaN   NaN       NaN  \
9200  21110  2022-02-01  2022-02-01        1049.00     +       1.0   
9201  21110  2022-01-31  2022-02-01       -1470.22     -       NaN   

       dt_tresor   dt_compta type_compta                          num_oper   
9199         NaN         NaN         NaN                               NaN  \
9200  2022-02-01  2022-02-01           C  8a3022ebea48b9cd08b6c6a4d75ea85a   
9201         NaN  2022-02-01         NaN  ef71a23f44ae162a64d68e77e2474198   

     cd_user cd_visa                   time  
9199     NaN     NaN  2022-02-02 00:00:00.0  
9200     NaN     NaN  2022-02-02 00:00:00.0  
9201     NaN     NaN  2022-02-02 00:00:00.0

In [58]:
# On va remplacer le lib_etendu NAN par unknown

data['lib_etendu'].fillna('unknown', inplace=True)

In [10]:
# On va etudier la colonne bilan

data['bilan'].value_counts()

bilan
21110    5951
21210    1167
47110     538
21140     332
44212     317
         ... 
21433       1
89412       1
89490       1
21446       1
46113       1
Name: count, Length: 80, dtype: int64

In [11]:
# Nombre de valeurs nulles dans la colonne bilan

data.bilan.isnull().sum()

0

In [12]:
# On va s'interesser a la colonne dt_val

data['dt_val'].value_counts()

dt_val
2022-02-01    3147
2022-01-31    2432
2022-01-28    1845
2022-02-02    1732
2022-02-16      76
              ... 
0353-03-31       1
5547-09-06       1
5546-01-05       1
5548-09-19       1
0284-02-29       1
Name: count, Length: 451, dtype: int64

In [34]:
# nombre de valeurs manquantes dans la colonne dt_val
data.dt_val.isnull().sum()

14

Conclusion :
On remarque qu'on a quelques valeurs nulles dans la colonne dt_val, mais qu'on a aussi des dates dont la plage depasse celle autorisee qui va du 1er janvier 1978 au 31 decembre 2262. On va donc remplacer toutes les dates qui sont hors de ces limites par le 1er janvier 1978. Puis remplacer les dates nulles par la date qui les precede

In [20]:
# Afficher les lignes qui ont une date nulle

data[data.dt_val.isnull()]

numero_compte                        lib_etendu   
994   77d1c7cdb8262110ad7ec1b24f56195f  e46f6c29f9c170baf9fbcc0d2255b422  \
995   77d1c7cdb8262110ad7ec1b24f56195f  2f70df074ad10784ad91e2494d799451   
996   77d1c7cdb8262110ad7ec1b24f56195f  1d92d261efdc6c51da47ee4c3f5ce51b   
998   77d1c7cdb8262110ad7ec1b24f56195f  c857737224ddc2ae29fe32f98a51420c   
999   77d1c7cdb8262110ad7ec1b24f56195f  0c5dde10055af6f3726ab005e4f1ed2a   
1600  0499067504948469b356fbe74e08d6d9  83f9c7bc5ab33c8a6c9a61a7cdeb57cf   
1601  0499067504948469b356fbe74e08d6d9  0c5dde10055af6f3726ab005e4f1ed2a   
3254  c507b3ee99facefc0aaa42e5b40fe9ab  0c5dde10055af6f3726ab005e4f1ed2a   
4120  6f93a70908558a8abbec920dabc3e29b  83f9c7bc5ab33c8a6c9a61a7cdeb57cf   
4125  6f93a70908558a8abbec920dabc3e29b  0c5dde10055af6f3726ab005e4f1ed2a   
5263  2964c143f717d9b26d4a9bde68b143b9  0c5dde10055af6f3726ab005e4f1ed2a   
6412  35604a4ecff5b4c8af905451c17a7fbe  0c5dde10055af6f3726ab005e4f1ed2a   
8748  021a9a2d47e6e317191634fb783ec9a3  0c5dde10055af6f3726ab005e4f1ed2a   
9199  5c01e4e913a1ba29ed1c2569e4373553                           unknown   

      bilan dt_val  montant_signe signe   dt_tresor   
994   47110    NaN         152.55     +         NaN  \
995   47110    NaN        1500.00     +  2022-02-01   
996   47110    NaN         163.31     +  2022-02-01   
998   47110    NaN           0.50     +  2022-02-01   
999   47110    NaN     5351219.71     +  2022-01-31   
1600  47110    NaN         572.48     +         NaN   
1601  47110    NaN     8179527.61     +  2022-01-31   
3254  47110    NaN    -3045953.35     -  2022-01-31   
4120  47110    NaN         572.48     +         NaN   
4125  47110    NaN      735215.99     +  2022-01-31   
5263  47110    NaN     1667557.87     +  2022-01-31   
6412  47110    NaN      656049.87     +  2022-01-31   
8748  47110    NaN     -800253.49     -  2022-01-31   
9199  21110    NaN            NaN   NaN         NaN   

                              num_oper cd_user cd_visa  
994                                NaN     NaN     NaN  
995   4ea543c9c9d17d90acd60b58cf877ebf     NaN     NaN  
996   4ea543c9c9d17d90acd60b58cf877ebf     NaN     NaN  
998   4ea543c9c9d17d90acd60b58cf877ebf     NaN     NaN  
999                                NaN     NaN     NaN  
1600                               NaN     NaN     NaN  
1601                               NaN     NaN     NaN  
3254                               NaN     NaN     NaN  
4120                               NaN     NaN     NaN  
4125                               NaN     NaN     NaN  
5263                               NaN     NaN     NaN  
6412                               NaN     NaN     NaN  
8748                               NaN     NaN     NaN  
9199                               NaN     NaN     NaN

In [15]:
data[data.dt_val.isnull()]['numero_compte'].value_counts()

numero_compte
77d1c7cdb8262110ad7ec1b24f56195f    5
0499067504948469b356fbe74e08d6d9    2
6f93a70908558a8abbec920dabc3e29b    2
c507b3ee99facefc0aaa42e5b40fe9ab    1
2964c143f717d9b26d4a9bde68b143b9    1
35604a4ecff5b4c8af905451c17a7fbe    1
021a9a2d47e6e317191634fb783ec9a3    1
5c01e4e913a1ba29ed1c2569e4373553    1
Name: count, dtype: int64

In [59]:
# On va remplacer les dates nulles par la date qui les precede

data['dt_val'].fillna(method='ffill', inplace=True)

In [60]:
# On va rempplacer les dates qui ne sont pas entre 2020 et 2023 par le 1er Janvier 2024
from datetime import datetime

date_str = "2024-01-01"
format_str = "%Y-%m-%d"
data['dt_val'] = data['dt_val'].apply(lambda x: date_str if x < '01/01/2020' or x > '31/12/2023' else x)

In [61]:
data['dt_val'].value_counts()

dt_val
2022-02-01    3148
2022-01-31    2432
2022-01-28    1846
2022-02-02    1732
2024-01-01     400
              ... 
0200-03-01       1
1007-02-28       1
0256-09-30       1
0514-08-31       1
0284-02-29       1
Name: count, Length: 135, dtype: int64

In [62]:
# On va remplacer toutes les dates qui ne sont pas de l'annee 2022 par le 1er Janvier 2024

data['dt_val'] = data['dt_val'].apply(lambda x: datetime.strptime(date_str, format_str) if datetime.strptime(x, format_str).year != 2022 else x)

In [67]:
# On va creer un colonne qui represente la difference en nombre de jours entre la colonne dt_val et dt_ope

data['diff_ope_val'] = (pd.to_datetime(data['dt_ope']) - pd.to_datetime(data['dt_val'])).dt.days

In [68]:
data['diff_ope_val'].value_counts()

diff_ope_val
 0.0      3148
 1.0      2432
 4.0      1845
-1.0      1732
-699.0     552
-15.0       76
 5.0        59
 2.0        53
 3.0        36
-3.0        25
 31.0       10
 6.0         9
 15.0        7
 26.0        4
-2.0         4
 7.0         3
 9.0         2
-14.0        1
 8.0         1
Name: count, dtype: int64

In [69]:
# On va etudier la colonne dt_tresor

data['dt_tresor'].value_counts()

dt_tresor
2022-02-01    5455
2022-01-31      66
Name: count, dtype: int64

In [70]:
# Valeurs nulles dans la colonne dt_tresor

data.dt_tresor.isnull().sum()

4479

In [71]:
# On va etudier la colonne num_oper

data['num_oper'].value_counts()

num_oper
33b740d576137674827e2107defdabe5    80
4ea543c9c9d17d90acd60b58cf877ebf    70
6b9dc969c1826a6ac3b52b53aef6ac16    12
f9ef806ecad58a628a68d8363c6e4837    12
414ac732be573e116062657474189e1f    11
                                    ..
54e271543ad2da7e6b67ee2173f912b9     1
bf4654d19ac164fd6c105105186c4a6e     1
8fd082f0c6fe2f078c89764b0004175d     1
116934cf4e39d76b0b13ab30fe97fcc3     1
14723d8b9a77ae0816b4943aa30bc8fe     1
Name: count, Length: 7751, dtype: int64

In [72]:
# Nombre de valeures nulles dans la colonne num_oper

data.num_oper.isnull().sum()

1401

In [73]:
# On va remplacer les valeurs nulles dans la colonne num_oper par unknown

data['num_oper'].fillna('unknown', inplace=True)

In [74]:
# On va etudier la colonne cd_user

data['cd_user'].value_counts()

cd_user
0346dc62eabe7a286bc2078850f06e76    519
f30f89aa344c312bd57bab8cc4a405f7    300
40c3846eac12f4696f2fdecd8ded0ee6    158
1ea352c2bd519e32e5ccf69e49898ca1    145
31e65299aa7f89cde760d5320f4ec1e6    143
bf8034a86c2eb3dd4d46dbc3330cf0ac    111
6c76dc5f5e0cd1f3ef07e9c1aff67837    110
98d6e6e2aac5ece115b222576619947f     93
5db7f6e1eed86658db53be6457059ac6     93
1d202445bd84e80678582ac28fbb46e6     90
e52352200d1828a8633816002778b9e5     79
899bfaa3587fd21f7f5445c5df05c6fc     79
c6226cab28e09b51aaab8c1071cd6790     76
6cfa0d55103cbf703390f15687c022b2     67
f6d068743d0ed6c85e58565a334bdde2     62
db36569c92ef37d31ffd032f094a8e27     58
8ee297eb54d847dcec04e03ac33013bf     53
1502bea97dc156e842d6a2a3fcefd9b3     52
73f4b91f9d8c3307c4b08df190a2138d     51
e796ea3f333d4732ccbf8c0cb05f87b2     50
e2a47310a7a53ff0196a5fc2b5eaae3d     50
60f7ed215938b6f2872e8f911e1c5844     46
6c9ef2063f2724a1c107d99cf0b25091     45
80388e81bba3d1703c3d0353a766dad4     42
858f6ac21a3fbe3a1a7fbf5cc565e621

In [75]:
# Nombre de valeurs nulles dans la colonne cd_user

data.cd_user.isnull().sum()

7326

In [76]:
# On va etudier cd_visa

data['cd_visa'].value_counts()

cd_visa
f30f89aa344c312bd57bab8cc4a405f7    300
31e65299aa7f89cde760d5320f4ec1e6    143
8a870234fe8538dadf538287d0aa313c     46
907f9180432dd9ca74d6d8ab0105ebc1     40
716e40954dd6035190f5a2b9285ef3d2     34
fd2a4efe69a4def6a615bdce39b0e7d0     22
28102905291e9b79f6ee7858d474c25e     22
03f8e8436282c26415a6318162e3b19e     22
66b3c4471795a5527c5faa319096c9ed     21
8ab0b016bbaa54c0ced1687e683521c2     17
bf941157511fb0e364510c014670cda9     11
c2a458d4fd78b85807d1ae4347999410      7
2342499d5c6861aec4addd3c3d24a9ec      7
656f0a452b15173a4208e628efc385b5      6
98d6e6e2aac5ece115b222576619947f      6
179f12d330bd75a2a191a1a27d416bc7      4
22b8d6696635356d5a91373559cf07fb      2
35808b2f1eddba8397e14cb92ab30913      2
34a5421af41446097ade5032d5bc7c28      2
700e15e8a37a0bf7e2be49b9c4078825      2
e2a47310a7a53ff0196a5fc2b5eaae3d      2
779a6676f26c07fe120e84c068b83d8c      2
e201f5c05116f04cd265e68919263c70      2
5d69a8f1b300b5ebdc8154ba0a206895      2
06fbaaa5f733bf1f6eac49310afdff1b

In [77]:
# Nombre de valeurs nulles dans la colonne cd_visa

data.cd_visa.isnull().sum()

9266

Suppression des colonnes inutiles

In [78]:
# On va retirer la colonne cpt_anal qui ne contient que des valeurs nulles car elle ne nous sera en aucun cas utile
data.drop('cpt_anal', axis=1, inplace=True)

# On va supprimmer les colonne dt_ope, extourne, dt_compta, type_compta, time qui ne contiennent que des valeurs uniques
data.drop(['dt_val', 'dt_ope', 'extourne', 'dt_compta', 'type_compta', 'dt_tresor', 'cd_user', 'cd_visa', 'time'], axis=1, inplace=True)

In [98]:
data.head()

lib_etendu  bilan  montant_signe signe  num_oper  diff_ope_val
0        1362  21110          -99.0     -      1936           4.0
1        5224  21110         -200.0     -      3318           1.0
2        4241  21110         1900.0     +      7751          -1.0
3         678  21110        -1000.0     -      2447           4.0
4         753  21110        -2700.0     -      7223           4.0

In [90]:
# Voyons les colonnes qui ont encore des valeurs nulles

data.isnull().sum()

lib_etendu       0
bilan            0
montant_signe    1
signe            1
num_oper         0
diff_ope_val     1
dtype: int64

In [100]:
# On va encoder les colonnes lib_etendu et num_oper, et signe

from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
data['lib_etendu'] = le.fit_transform(data['lib_etendu'])
data['num_oper'] = le.fit_transform(data['num_oper'])

In [3]:
data = pd.read_csv('Outputs/data.csv')

In [5]:
print(len(data.loc[(data['signe'] == 1) & (data['montant_signe'] > 0)]))

0


In [6]:
# remplacer les valeurs nulles de la colonne montant_signe par la valeur la plus frequente

data['montant_signe'].fillna(data['montant_signe'].mode()[0], inplace=True)

In [7]:
# remplacer les valeurs nulles de la colonne diff_ope_val par la valeur la plus frequente

data['diff_ope_val'].fillna(data['diff_ope_val'].mode()[0], inplace=True)

In [8]:
# On va maintenant sauvegarder notre dataset pour l'utiliser pour faire le benchmark

data.to_csv('Outputs/data.csv', index=False)